types/hast/index.d.ts

```diff
@@ -10,14 +10,62 @@ import { Parent as UnistParent, Literal as UnistLiteral, Node as UnistNode } fro
 
 export { UnistNode as Node };
 
+/**
+ * This map registers all node types that may be used as top-level content in the document.
+ *
+ * These types are accepted inside `root` nodes.
+ *
+ * This interface can be augmented to register custom node types.
+ *
+ * @example
+ * declare module 'hast' {
+ *   interface RootContentMap {
+ *     // Allow using raw nodes defined by `rehype-raw`.
+ *     raw: Raw;
+ *   }
+ * }
+ */
+export interface RootContentMap {
+    comment: Comment;
+    doctype: DocType;
+    element: Element;
+    text: Text;
+}
+
+/**
+ * This map registers all node types that may be used as content in an element.
+ *
+ * These types are accepted inside `element` nodes.
+ *
+ * This interface can be augmented to register custom node types.
+ *
+ * @example
+ * declare module 'hast' {
+ *   interface RootContentMap {
+ *     custom: Custom;
+ *   }
+ * }
+ */
+export interface ElementContentMap {
+    comment: Comment;
+    element: Element;
+    text: Text;
+}
+
+export type Content = RootContent | ElementContent;
+
+export type RootContent = RootContentMap[keyof RootContentMap];
+
+export type ElementContent = ElementContentMap[keyof ElementContentMap];
```

Comment:
I don’t necessarily think we need to define the types `Content`, `RootContent`, and `ElementContent`. Their implementation could be directly used when defining children.

```ts
export interface Parent extends UnistParent {
    /**
     * List representing the children of a node.
     */
    children: Array<RootContentMap[keyof RootContentMap] | ElementContentMap[keyof ElementContentMap]>;
}
```

I kept this approach in `mdast`, because those types already existed and I didn’t want to make breaking changes.

If you think this approach is better, we can keep it though.

Reply:
I relatively frequently have to access them and have been doing it like this: https://github.com/rehypejs/rehype-external-links/blob/ae637bebd5ed30a1f931cb522b87e58a728da2f2/index.js#L4.
Either works (`ElementContentMap[keyof ElementContentMap][]` or `Element['children'][number]`), but I feel `ElementContent` would be cleaner.

Do you sometimes access these content models in your types?

Reply:
> Do you sometimes access these content models in your types? And if so, what do you use?
So far: no. I prefer to write code in a way TypeScript can infer these types. But I can see why this is useful sometimes. I think the linked code is a good example of why these types would be useful to keep.